### PR TITLE
Improve HA code

### PR DESCRIPTION
**Pull request by as51340**
### Description
Changes made:

Renamed `ReplicationInstance` -> `ReplicationInstanceConnector`
Renamed`CoordinatorClient` -> `ReplicationInstanceClient`
Renamed `coordinator_server_id` -> `coordinator_id`
When created, coordinator instance (self) is added to Raft log.
`ReplicationInstanceClient` is injected into `ReplicationInstanceConnector` to enable mocking.
Fixed the bug which forbade creating Memgraph instance in enterprise version without specifying flags used for high availability.
Fixed NuRaft's bug when destructing raft_server occurring because of `launcher_.shutdown()`.
Unit tested `CoordinatorClusterState`.
Unit tested `RaftState`.
Unit tested `CoordinatorStateMachine`.
Unit tested `CoordinatorInstance`.
Mocked `ReplicationInstanceClient`.

[master < Task] PR
- [x] Provide the full content or a guide for the final git message
    - **Improve HA code**

### Documentation checklist
- [x] Add the documentation label tag
- [x] Add the bug / feature label tag
- [x] Add the milestone for which this feature is intended
    - If not known, set for a later milestone
- [x] Write a release note, including added/changed clauses
    - **[Release note text]**
- [ ] Link the documentation PR here
    - **[Documentation PR link]**
- [x] Tag someone from docs team in the comments

### CODE REVIEW

**Comment by as51340** [comment]
@kgolubic 2 release notes:
1. Fixed bug when shutting down Raft server
2. Allow initializing Memgraph enterprise instance without specifying any flag used in high-availability

File: src/coordination/raft_state.cpp [hunk 1]
```diff
@@ -426,5 +432,50 @@ auto RaftState::GetCoordinatorInstances() const -> std::vector<CoordinatorInstan
...
 }
 
+auto RaftState::GetRoutingTable() const -> RoutingTable {
+  auto res = RoutingTable{};
+
+  auto const repl_instance_to_bolt = [](ReplicationInstanceState const &instance) {
+    return instance.config.BoltSocketAddress();
+  };
+
+  // TODO: (andi) This is wrong check, Fico will correct in #1819.
```
[the remark this answers is not in the document]
Comment (as51340) [reply src/coordination/raft_state.cpp:442]: I don't know how you intended to fix it, add it in force reset

File: src/coordination/coordinator_cluster_state.cpp [hunk 2]
```diff
@@ -29,9 +29,29 @@ void from_json(nlohmann::json const &j, ReplicationInstanceState &instance_state
...
-    : repl_instances_{std::move(instances)}, current_main_uuid_(current_main_uuid), is_lock_opened_(is_lock_opened) {}
+    : repl_instances_{std::move(instances)},
+      coordinators_{std::move(coordinators)},
+      current_main_uuid_(current_main_uuid),
+      is_lock_opened_(is_lock_opened) {}
+
+CoordinatorClusterState::CoordinatorClusterState(CoordinatorInstanceInitConfig const &config) {
+  auto c2c_config = CoordinatorToCoordinatorConfig{
+      .coordinator_id = config.coordinator_id,
+      .bolt_server = io::network::Endpoint{"127.0.0.1", static_cast<uint16_t>(config.bolt_port)},
```
[the remark this answers is not in the document]
Comment (as51340) [reply src/coordination/coordinator_cluster_state.cpp:51]: I don't think this is correct but I don't think Memgraph has any capability of using some other network parameter

[the remark this answers is not in the document]
Comment (as51340) [reply src/coordination/coordinator_cluster_state.cpp:51]: I think replication is using 0.0.0.0 which means listen on every available network interface.

File: src/coordination/raft_state.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (antoniofilipovic) [reply src/coordination/raft_state.cpp:406]: Aha, now you have functions called HasMainState and IsCurrentMain. To get the current main you can use IsCurrentMain(instance_name). HasMainState returns true if status is MAIN, but that doesn't mean that instance is MAIN. IsCurrentMain checks also for instance_uuid

**Comment by kgolubic** [comment]
Relase note entry:

```
Fixed a bug that could occur when shutting down the Raft server.
Initializing Memgraph enterprise instance without specifying any flag used in high-availability is now allowed.
```

**Comment by as51340** [comment]
yes here it will be added https://github.com/memgraph/documentation/pull/649

**Pull request merged**